src/Shared/MSBuildLoadContext.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
+using System;
+using System.Collections.Immutable;
+using System.Reflection;
+using System.Runtime.Loader;
+
+#nullable enable
+namespace Microsoft.Build.Shared
+{
+    /// <summary>
+    /// This class is used to isolate the types used by an MSBuild plugin
+    /// (SDK resolver, logger, or task).
+    /// </summary>
+    internal class MSBuildLoadContext : AssemblyLoadContext
+    {
+        private readonly AssemblyDependencyResolver _resolver;
+        private readonly string _directory;
+
+        private static readonly ImmutableHashSet<string> _wellKnownAssemblyNames =
+            new[]
+            {
+                "MSBuild",
+                "Microsoft.Build",
+                "Microsoft.Build.Framework",
+                "Microsoft.Build.Tasks.Core",
+                "Microsoft.Build.Utilities.Core",
+            }.ToImmutableHashSet();
+
+
+        public MSBuildLoadContext(string assemblyPath) :
+            base(name: assemblyPath, isCollectible: false) // TODO: make this collectible?
+        {
+            _directory = assemblyPath;
+            _resolver = new AssemblyDependencyResolver(_directory);
```

Comment:
I just want to make sure this is intentional - this will look for `task.deps.json` next to the `task.dll` and if it's there it will use it - new and potentially breaking behavior.
If it's not there it will use all `.dll` files in the same folder as the dependencies - which is probably closer to the current behavior.

I'm not saying we should not do this - in fact not using the `.deps.json` will basically fix only half of the problem we could fix with this change. I'm just pointing out that this is potentially a breaking change.

Reply:
Talked about this with @livarcocc. We're planning to be conservative now (use the same behavior but in an ALC) and reconsider adding an `AssemblyDependencyResolver` later (probably in the .NET 5 timeframe). We agree that it's better moving forward but don't want to break folks if possible.

Reply:
OK - this means that tasks with native (or platform specific dependencies) will still not be easy to write. For example SourceLink is such a task - where it needs to implement a two-layer approach and implement it's own ALC and custom resolution and so on. Such tasks carry some of their dependencies in subfolders which are keyed by the platform - so for example `win/NativeLib.dll` and `linux/NativeLib.so` - similarly for managed dependencies which are different per-platform. Using `AssemblyDependencyResolver` in some way would make this possible as it will be able to correctly pick the right one. The current solution will not help with these tasks.

I'm writing this down just to make it clear, I fully understand the desire to keep this as risk-free as possible.

There might be some way to combine both `AssemblyDependencyResolver` and plain directory search together to achieve a middle ground - but it would require .NET Core 3.0, so we're not there yet anyway.

Reply:
Filed microsoft/msbuild#5037 to make sure we come back to reevaluate this.